Content.Shared/_CM14/Xenos/Charge/SharedXenoChargeSystem.cs

```diff
@@ -0,0 +1,141 @@
+            return;
+
+        if (!_xenoPlasma.TryRemovePlasmaPopup(xeno.Owner, xeno.Comp.PlasmaCost))
+            return;
+
+        args.Handled = true;
+
+        var ev = new XenoChargeDoAfterEvent(GetNetCoordinates(args.Target));
+        var doAfter = new DoAfterArgs(EntityManager, xeno, xeno.Comp.ChargeDelay, ev, xeno)
+        {
+            BreakOnMove = true,
+            Hidden = true,
+        };
+
+        _stun.TrySlowdown(xeno, TimeSpan.FromSeconds(1.75f), false, 0f, 0f);
+        _doAfter.TryStartDoAfter(doAfter);
+    }
+
+    private void OnXenoChargeDoAfterEvent(Entity<XenoChargeComponent> xeno, ref XenoChargeDoAfterEvent args)
+    {
+        if (args.Cancelled)
+            return;
+
+        var coordinates = GetCoordinates(args.Coordinates);
+        var origin = _transform.GetMapCoordinates(xeno);
+        var diff = coordinates.ToMap(EntityManager, _transform).Position - origin.Position;
+        var length = diff.Length();
+        diff *= xeno.Comp.Range / length;
+
+        xeno.Comp.Charge = diff;
+        Dirty(xeno);
+
+        _throwing.TryThrow(xeno, diff, 10, animated: false);
+    }
+
+    private void OnXenoChargeHit(Entity<XenoChargeComponent> xeno, ref ThrowDoHitEvent args)
+    {
+        // TODO CM14 lag compensation
+        var targetId = args.Target;
+        if (_physicsQuery.TryGetComponent(xeno, out var physics) &&
+            _thrownItemQuery.TryGetComponent(xeno, out var thrown))
+        {
+            _thrownItem.LandComponent(xeno, thrown, physics, true);
+            _thrownItem.StopThrow(xeno, thrown);
+        }
+
+        if (_timing.IsFirstTimePredicted && xeno.Comp.Charge is { } charge)
+        {
+            xeno.Comp.Charge = null;
+            DoLunge(xeno, charge.Normalized());
+        }
+
+        if (TryComp(xeno, out XenoComponent? xenoComp) &&
+            TryComp(targetId, out XenoComponent? targetXeno) &&
+            xenoComp.Hive == targetXeno.Hive)
+        {
+            return;
+        }
+
+        if (_net.IsServer)
+            _audio.PlayPvs(xeno.Comp.Sound, xeno);
+
+        var damage = _damageable.TryChangeDamage(targetId, xeno.Comp.Damage);
+        if (damage?.GetTotal() > FixedPoint2.Zero)
+        {
+            var filter = Filter.Pvs(targetId, entityManager: EntityManager).RemoveWhereAttachedEntity(o => o == xeno.Owner);
+            _colorFlash.RaiseEffect(Color.Red, new List<EntityUid> { targetId }, filter);
+        }
+
+        var origin = _transform.GetMapCoordinates(xeno);
+        var target = _transform.GetMapCoordinates(targetId);
+        var diff = target.Position - origin.Position;
+        var length = diff.Length();
+        diff *= xeno.Comp.Range / 3 / length;
+
+        _stun.TryParalyze(targetId, xeno.Comp.StunTime, true);
+        _throwing.TryThrow(targetId, diff, 10);
+
+        if (_net.IsServer &&
+            HasComp<MarineComponent>(targetId))
+            SpawnAttachedTo(xeno.Comp.Effect, targetId.ToCoordinates());
+    }
```

Comment:
The code here is almost the same as the one in HeadbuttSystem, this code should be in a separate method that both call.